src/compiler/checker.ts

```diff
@@ -8773,26 +8819,135 @@ module ts {
+         */
+        function getIteratedType(iterable: Type, expressionForError: Expression): Type {
+            Debug.assert(languageVersion >= ScriptTarget.ES6);
+            var iteratedType = getIteratedTypeSubroutine(iterable, expressionForError);
+            // Now even though we have extracted the iteratedType, we will have to validate that the type
+            // passed in is actually an Iterable.
+            if (expressionForError && iteratedType) {
+                var completeIterableType = globalIterableType !== emptyObjectType ? createTypeReference(<GenericType>globalIterableType, [iteratedType]) : emptyObjectType;
+                checkTypeAssignableTo(iterable, completeIterableType, expressionForError);
+            }
+
+            return iteratedType;
+            
+            function getIteratedTypeSubroutine(iterable: Type, expressionForError: Expression) {
+                if (allConstituentTypesHaveKind(iterable, TypeFlags.Any)) {
+                    return undefined;
+                }
+
+                // We want to treat type as an iterable, and get the type it is an iterable of. The iterable
+                // must have the following structure (annotated with the names of the variables below):
+                //
+                // { // iterable
+                //     [Symbol.iterator]: { // iteratorFunction
+                //         (): { // iterator
+                //             next: { // iteratorNextFunction
+                //                 (): { // iteratorNextResult
+                //                     value: T // iteratorNextValue
+                //                 }
+                //             }
+                //         }
+                //     }
+                // }
+                //
+                // T is the type we are after. At every level that involves analyzing return types
+                // of signatures, we union the return types of all the signatures.
+                var iteratorFunction = getTypeOfPropertyOfType(iterable, getPropertyNameForKnownSymbolName("iterator"));
+                if (iteratorFunction && allConstituentTypesHaveKind(iteratorFunction, TypeFlags.Any)) {
+                    return undefined;
+                }
+
+                var iteratorFunctionSignatures = iteratorFunction ? getSignaturesOfType(iteratorFunction, SignatureKind.Call) : emptyArray;
+                if (iteratorFunctionSignatures.length === 0) {
+                    if (expressionForError) {
+                        error(expressionForError, Diagnostics.The_right_hand_side_of_a_for_of_statement_must_have_a_Symbol_iterator_method_that_returns_an_iterator);
+                    }
+                    return undefined;
+                }
+
+                var iterator = getUnionType(map(iteratorFunctionSignatures, getReturnTypeOfSignature));
+                if (allConstituentTypesHaveKind(iterator, TypeFlags.Any)) {

```

Comment:
This function is essentially generating a semantic access chain of the following form (given the type `iterable`):
```ts
iterable[Symbol.iterator]().next().value
```
At every level in this access chain, you could encounter an `anyType` or `unknownType`. So let's say `iterable[Symbol.iterator]()` returns `anyType`. Then when we try to access `next`, we will get undefined, because `anyType` does not have properties. Now consider what would happen if we did not do the `any` check. We would proceed, and `iteratorNextFunction` would be undefined. Then `iteratorNextFunctionSignatures` would be the emptyArray. And then we would incorrectly report an error that you don't have a next method.